File: SSolver/Random.fs
//--------------------------------------------------------------------------------
// This file is part of The Soul, A Neural Network Simulation System.
//
// Copyright © 2010 LBE Group. All rights reserved.
//
// For information about this application and licensing, go to http://soul.codeplex.com
//
// THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
// EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
//--------------------------------------------------------------------------------

namespace SSolver

open System

/// <summary>
/// Random Number Generator based on System.Random
/// </summary>
type RNG = 
    inherit Random
    new() = {inherit Random()}
    new(seed:int) = {inherit Random(seed)}
    override this.Sample() = 
        base.Sample()
    override this.NextDouble() = 
        this.Sample()
    override this.Next() = 
        this.Next(Int32.MaxValue)
    override this.Next(maxValue:int) = 
        int (this.Sample() * float maxValue)
    override this.Next(minValue:int,maxValue:int) = 
        if minValue > maxValue then
            raise(new ArgumentOutOfRangeException("minValue","ArgumentMinValueGreaterThanMaxValue"))
        int (this.Sample() * float(maxValue - minValue)) + minValue
    abstract Nexts: int*int->int[]
    default this.Nexts(n:int,maxValue:int) = 
        if n<1 then
            raise(new ArgumentException("ArgumentMustBePositive"))
        Array.init n (fun x ->this.Next(maxValue))
    override this.NextBytes(buffer:byte[]) =
        if buffer = null then
            raise(new ArgumentNullException("buffer"))
        for i=0 to buffer.Length-1 do
            buffer.[i] <- byte( this.Next() % 256 )
        ()
    abstract NextDoubles: int->float[]
    default this.NextDoubles(n:int) = 
        if n<1 then
            raise(new ArgumentException("ArgumentMustBePositive"))
        Array.init n (fun x ->this.NextDouble())
    abstract NextMeanStdDouble: float*float->float
    default this.NextMeanStdDouble(mean:float,std:float) = 
        let minValue = mean - std
        let maxValue  = mean + std
        this.Sample() * (maxValue - minValue) + minValue

/// <summary>
/// Generate Gaussian Random Number Using Polar Form of Box-Muller Transformation
/// </summary>
type GaussRNG = 
    inherit RNG
    val mutable private gn:float
    val mutable private ign:int
    val mutable Random:RNG
    new() = new GaussRNG(new RNG())
    new(seed:int)=new GaussRNG(new RNG(seed))
    new(random:RNG)={gn=0.0;ign=0;Random = random}
    /// <summary>
    /// Gets Standard Gaussian Distribution Number (mean: 0, std: 1)
    /// </summary>
    /// <returns></returns>
    override this.Sample() = 
        let mutable un1=2.0 * this.Random.NextDouble() - 1.0
        let mutable un2 = 2.0 * this.Random.NextDouble() - 1.0
        let mutable w = un1*un1 + un2*un2
        if this.ign=0 then
            while (w >= 1.0 || w = 0.0) do
                un1<-2.0 * this.Random.NextDouble() - 1.0
                un2 <- 2.0 * this.Random.NextDouble() - 1.0
                w <- un1*un1 + un2*un2
            let sqw = Math.Sqrt((-2.0 * Math.Log(w)) / w)
            this.gn <- un2 * sqw
            this.ign <- 1
            un1 * sqw
        else
            this.ign <- 0
            this.gn
    /// <summary>
    /// Gets Custom Gaussian Distribution Number (mean, std)
    /// </summary>
    /// <param name="mean"></param>
    /// <param name="std"></param>
    /// <returns></returns>
    override this.NextMeanStdDouble(mean:float,std:float) = 
        CoreFunc.RNMeanStd (this.Sample()) mean std 

File: SCore/Random.cs
//--------------------------------------------------------------------------------
// This file is part of The Soul, A Neural Network Simulation System.
//
// Copyright © 2010 LBE Group. All rights reserved.
//
// For information about this application and licensing, go to http://soul.codeplex.com
//
// THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
// EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
//--------------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using SSolver;
using System.Reflection;
using System.Windows.Media.Media3D;

namespace SCore
{
    public interface IRandomizable
    {
        void Randomize(bool isincludechildren = false, params Tuple<string, Randomizer>[] targets);
        void Randomize(string property, Randomizer randomizer, bool isincludechildren = false);
    }

    public class Randomizer
    {
        int dimxbegin, dimxend, dimybegin, dimyend, dimzbegin, dimzend;
        RNG randomsource;
        double mean, std;


        public Randomizer(RNG randomsource, int dimxbegin = 0, int dimxend = 0, int dimybegin = 0, int dimyend = 0, int dimzbegin = 0, int dimzend = 0, double mean = 0.5, double std = 0.5)
        {
            this.randomsource = randomsource;
            this.dimxbegin = dimxbegin;
            this.dimxend = dimxend;
            this.dimybegin = dimybegin;
            this.dimyend = dimyend;
            this.dimzbegin = dimzbegin;
            this.dimzend = dimzend;
            this.mean = mean;
            this.std = std;
        }


        public virtual void Randomize(object[, ,] dimensionelements, PropertyInfo propertyinfo)
        {
            if (randomsource != null)
            {
                PrepareDimensionRange(dimensionelements.GetLength(0), dimensionelements.GetLength(1), dimensionelements.GetLength(2));
                try
                {
                    bool isneedconvert = propertyinfo.PropertyType != typeof(double);
                    for (var i = dimxbegin; i <= dimxend; i++)
                    {
                        for (var j = dimybegin; j <= dimyend; j++)
                        {
                            for (var k = dimzbegin; k <= dimzend; k++)
                            {
                                var randomvalue = randomsource.NextMeanStdDouble(mean, std);
                                propertyinfo.SetValue(dimensionelements[i, j, k], Convert(isneedconvert, randomvalue, propertyinfo.PropertyType), null);
                            }
                        }
                    }
                }
                catch (Exception e)
                {
                }
            }
        }

        protected virtual void PrepareDimensionRange(int dimxlength, int dimylength, int dimzlength)
        {
            dimxbegin = Math.Min(dimxbegin, dimxlength - 1);
            dimxend = Math.Min(dimxend, dimxlength - 1);
            if (dimxbegin > dimxend)
            {
                throw new ArgumentOutOfRangeException("DimXBeginValue", "ArgumentBeginValueGreaterThanEndValue");
            }
            dimybegin = Math.Min(dimybegin, dimylength - 1);
            dimyend = Math.Min(dimyend, dimylength - 1);
            if (dimybegin > dimyend)
            {
                throw new ArgumentOutOfRangeException("DimYBeginValue", "ArgumentBeginValueGreaterThanEndValue");
            }
            dimzbegin = Math.Min(dimzbegin, dimzlength - 1);
            dimzend = Math.Min(dimzend, dimzlength - 1);
            if (dimzbegin > dimzend)
            {
                throw new ArgumentOutOfRangeException("DimZBeginValue", "ArgumentBeginValueGreaterThanEndValue");
            }
        }

        protected virtual object Convert(bool isconvert, double value, Type outtype)
        {
            if (isconvert)
            {
                if (outtype == typeof(Point3D))
                {
                    return new Point3D(value, randomsource.NextMeanStdDouble(mean, std), randomsource.NextMeanStdDouble(mean, std));
                }
            }
            return value;
        }

    }

    public static class Poisson
    {
        private static Random ran;

        static Poisson()
        {
            ran = new Random();
        }

        /// <summary>
        /// 负指数分布随机数产生
        /// </summary>
        /// <param name="lam">参数</param>
        /// <returns></returns>
        public static double ngtIndex(double lam)
        {
            double dec = ran.NextDouble();
            while (dec == 0)
                dec = ran.NextDouble();
            return -Math.Log(dec) / lam;
        }

        /// <summary>
        /// 泊松分布产生
        /// </summary>
        /// <param name="lam">参数</param>
        /// <param name="time">时间</param>
        /// <returns></returns>
        public static double poisson(double lam, double time)
        {
            int count = 0;

            while (true)
            {
                time -= ngtIndex(lam);
                if (time > 0)
                    count++;
                else
                    break;
            }
            return count;
        }
    }

}
